icicle/appUtils/msm/msm.cu

```diff
@@ -992,21 +799,22 @@ namespace msm {
   template <typename S, typename A, typename P>
   cudaError_t MSM(S* scalars, A* points, int msm_size, MSMConfig& config, P* results)
   {
-    int bitsize = (config.bitsize == 0) ? S::NBITS : config.bitsize;
+    const int bitsize = (config.bitsize == 0) ? S::NBITS : config.bitsize;
     cudaStream_t& stream = config.ctx.stream;
-    // TODO: DmytroTym/HadarIngonyama - unify the implementation of the bucket method and the batched bucket method in
-    // one function
-    if (config.batch_size == 1)
-      return CHK_STICKY(bucket_method_msm(
-        bitsize, 16, scalars, points, msm_size, results, config.are_scalars_on_device,
-        config.are_scalars_montgomery_form, config.are_points_on_device, config.are_points_montgomery_form,
-        config.are_results_on_device, config.is_big_triangle, config.large_bucket_factor, config.is_async, stream));
-    else
-      return CHK_STICKY(batched_bucket_method_msm(
-        bitsize, (config.c == 0) ? get_optimal_c<S>(msm_size) : config.c, scalars, points, config.batch_size, msm_size,
-        (config.points_size == 0) ? msm_size : config.points_size, results, config.are_scalars_on_device,
-        config.are_scalars_montgomery_form, config.are_points_on_device, config.are_points_montgomery_form,
-        config.are_results_on_device, config.is_async, stream));
+
+    unsigned c = config.batch_size > 1 ? ((config.c == 0) ? get_optimal_c(msm_size) : config.c) : 16;
+    // reduce c to closest power of two if not using big_triangle reduction logic
+    // TODO: support arbitrary values of c
+    if (!config.is_big_triangle) {
+      while ((c & (c - 1)) != 0)
```

Comment:
Can't this procedure just hang for some values of `c`? Like, if `c=b101`, then `c & (c - 1)` doesn't add any extra bits so the process is never terminated. Also since the idea is rounding up I'm afraid any MSM of size bigger than 2^20 will end up with `c=32` which takes way too much memory

Reply:
it will never hang and it rounds down, not up:
- for powers of two exactly 1 bit is 1'b1 --> meaning 'c & (c-1)' is zero. for example for c=4'b0100, c-1 = 4'b0011.
- otherwise each iteration is masking one bit to 1'b0, making the number smaller and therefore it must end in at most 31 iterations.
- note that masking c with any number cannot result in a number larger than c, meaning it is rounding down.

in your example c=b101 therefore c-1=b100 --> c& (c-1) = b100. Note that the first non zero bit is masked to 1'b0 and next iteration would mask the next non zero bit.

Reply:
oh sorry you're right